### PR TITLE
Merge bitcoinjs v0.6.1

### PR DESCRIPTION
**Pull request by louisinger**
This PR merges the changes of bitcoinjs v0.6 (see #36 for detailed changelog of bitcoinjs).

The major change is about `WitnessForHashV1` in `Transaction`class. _This is the first step for a full taproot support_. However, the way elements computes WitnessV1 hash is a bit different from Bitcoin, that's why this function differs a lot from the original one. `taproot.test.ts` let u receive and spent a single-key taproot coin (unconfidential only) but complex taproot script are not supported yet.

Other important (~breaking) changes from bitcoinjs v0.6:
* `psbt.validateSignatureOfInput` takes an additional **optional** argument: the `validator`.
* some types has been "merged"
* new getters in Psbt (`getInputs` and `getOutputs`)
* **bech32m** encoding support (unconfidential only)

I kept the same tests and fixtures as the master branch, Tx and Psbt encoding / decoding are not broken.

it closes #36 

EDIT: taproot
---
latest commits add `bip341.ts` filte, exporting useful function for a full taproot support. `taproot.spec.ts` has been modified to test key path spend and script path spend of a "taproot coin" (both unconfidential).

@tiero please review

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,50 @@
+# 6.0.0
```
Comment (tiero) [inline CHANGELOG.md]: Do we need this?

File: LICENSE [hunk 2]
```diff
@@ -1,6 +1,6 @@
 The MIT License (MIT)
 
-Copyright (c) 2011-2019 bitcoinjs-lib contributors
+Copyright (c) 2011-2020 bitcoinjs-lib contributors
```
Comment (tiero) [inline LICENSE]: I do think we shall add 
```suggestion
Copyright (c) 2019-2022 Vulpem Ventures
```

File: ts_src/crypto.ts [hunk 3]
```diff
@@ -31,3 +31,30 @@ export function hash160(buffer: Buffer): Buffer {
...
   return sha256(sha256(buffer));
 }
+
+const TAGS = [
+  'BIP0340/challenge',
+  'BIP0340/aux',
+  'BIP0340/nonce',
+  'TapLeaf',
+  'TapLeaf/elements',
+  'TapBranch',
```
Comment (tiero) [inline ts_src/crypto.ts]: We need `TapBranch/elements` maybe?

File: ts_src/bufferutils.ts [hunk 4]
```diff
@@ -87,7 +98,7 @@ export class BufferWriter {
     this.offset += slice.copy(this.buffer, this.offset);
   }
 
-  writeVarSlice(slice: Buffer): void {
+  writeVarSlice(slice: Buffer = Buffer.alloc(1)): void {
```
Comment (tiero) [inline ts_src/bufferutils.ts]: add a comment for this? can be unexpected if used outside

File: test/integration/taproot.spec.ts [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+} from '../../ts_src';
+import { broadcast, TESTNET_APIURL } from './_regtest';
+import {
+  confidentialValueToSatoshi,
+  satoshiToConfidentialValue,
+} from '../../ts_src/confidential';
+import { TestnetGenesisBlockHash } from '../../ts_src/transaction';
+const net = networks.testnet;
+
+describe('bitcoinjs-lib (transaction with taproot)', () => {
```
Comment (tiero) [inline test/integration/taproot.spec.ts]:
```suggestion
describe('liquidjs-lib (transaction with taproot)', () => {
```

File: CHANGELOG.md (hunk 1 above)
Comment (tiero) [reply CHANGELOG.md]: we can drop it

**Review by tiero — APPROVED** [review]
LGTM after we make sure we do not break current test suite (and apps that depends on old LiquidJS)

File: ts_src/bufferutils.ts (hunk 4 above)
Comment (louisinger) [reply ts_src/bufferutils.ts]: Finally, I've removed default value and handle undefined/null at upper level

File: src/networks.d.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+/// <reference types="node" />
+import { Network as BitcoinJSNetwork } from 'ecpair/src/networks';
+export declare type GenesisBlockHash = Buffer;
+export interface Network extends BitcoinJSNetwork {
+    blech32: string;
+    assetHash: string;
+    confidentialPrefix: number;
+    genesisBlockHash: GenesisBlockHash;
```
Comment (tiero) [inline src/networks.d.ts:8]: I would make it optional, to no break all existing Network definition

Comment (louisinger) [reply src/networks.d.ts:8]: network definition are often (always?) fetched from liquidjs isn't it? the new definitions includes the genesis block so it shouldn't break if the app updates liquidjs? Moreover, genesis hash are constants so even if someone defines a new regtest network, he could use the one exported via `regtest` (so easy to update)

**Comment by louisinger** [comment]
`f3314cf` improves the `bip341.ts` API, now it uses the one proposed by [bitcoinjs PR#1742 (taprootutils.ts)](https://github.com/bitcoinjs/bitcoinjs-lib/pull/1742#issue-1044094132).

**Pull request merged**